### PR TITLE
Factor out error reporting from `smart_resolve_path_fragment` fn

### PR DESCRIPTION
**Pull request by alexreg**
This function was ridiculously monolithic before. We now have three rather-less-monolithic-and-horrifying functions.

r? @centril

### CODE REVIEW

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -3123,391 +3123,405 @@ impl<'a> Resolver<'a> {
...
-                        format!("there is an enum variant `{}`{}; ",
-                                enum_candidates[0].0, others)
-                    } else {
-                        String::new()
-                    };
-                    let msg = format!("{}try using the variant's enum", preamble);
+        let mut levenshtein_worked = false;
 
-                    err.span_suggestions(
+        // Try Levenshtein algorithm.
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: From here (1)...

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -3123,391 +3123,405 @@ impl<'a> Resolver<'a> {
...
+            );
+            err.span_suggestion(
+                ident_span,
+                &msg,
+                suggestion.candidate.to_string(),
+                Applicability::MaybeIncorrect,
+            );
+
+            levenshtein_worked = true;
+        }
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: to here (1)... seems like a good place to break into a function.

File: src/librustc_resolve/lib.rs [hunk 3]
```diff
@@ -3123,391 +3123,405 @@ impl<'a> Resolver<'a> {
...
+
+        // Try context-dependent help if relaxed lookup didn't work.
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: From here (2)...

File: src/librustc_resolve/lib.rs [hunk 4]
```diff
@@ -3123,391 +3123,405 @@ impl<'a> Resolver<'a> {
...
+                              implemented struct");
+                    return (err, candidates);
                 }
+                (Def::TyAlias(_), _) | (Def::AssociatedTy(..), _) if ns == ValueNS => {
+                    err.note("can't use a type alias as a constructor");
+                    return (err, candidates);
+                }
+                _ => {}
             }
+        }
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: to here (2).. seems like a good place to break into a function also.

File: src/librustc_resolve/lib.rs [hunk 5]
```diff
@@ -3123,391 +3123,405 @@ impl<'a> Resolver<'a> {
...
-
-                __diagnostic_used!(E0424);
-                err.code(DiagnosticId::Error("E0424".into()));
-                err.span_label(span, match source {
-                    PathSource::Pat => {
-                        format!("`self` value is a keyword \
-                                and may not be bound to \
-                                variables or shadowed")
+        }
+        if path.len() == 1 && self.self_type_is_available(span) {
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: Here (3)...

File: src/librustc_resolve/lib.rs [hunk 6]
```diff
@@ -3123,391 +3123,405 @@ impl<'a> Resolver<'a> {
...
+                            span,
+                            "try",
+                            format!("Self::{}", path_str),
+                            Applicability::MachineApplicable,
+                        );
+                    }
+                }
+                return (err, candidates);
             }
+        }
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: To here (3)... might be another good place to split up.

File: src/librustc_resolve/lib.rs [hunk 7]
```diff
@@ -3123,391 +3123,405 @@ impl<'a> Resolver<'a> {
...
+                _ => {
+                    format!("`self` value is a keyword \
+                             only available in methods \
+                             with `self` parameter")
+                }
+            });
+            return (err, Vec::new());
+        }
+
+        // Try to lookup name in more relaxed fashion for better error reporting.
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: Somewhere here (4)...

File: src/librustc_resolve/lib.rs [hunk 8]
```diff
@@ -3123,391 +3123,405 @@ impl<'a> Resolver<'a> {
...
-                debug!("smart_resolve_path_fragment E0424 source:{:?}", source);
-
-                __diagnostic_used!(E0424);
-                err.code(DiagnosticId::Error("E0424".into()));
-                err.span_label(span, match source {
-                    PathSource::Pat => {
-                        format!("`self` value is a keyword \
-                                and may not be bound to \
-                                variables or shadowed")
+        }
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: To here (4)..

File: src/librustc_resolve/lib.rs [hunk 9]
```diff
@@ -3123,391 +3123,405 @@ impl<'a> Resolver<'a> {
...
+
+        // Emit special messages for unresolved `Self` and `self`.
+        if is_self_type(path, ns) {
+            __diagnostic_used!(E0411);
+            err.code(DiagnosticId::Error("E0411".into()));
+            err.span_label(span, format!("`Self` is only available in impls, traits, \
+                                          and type definitions"));
+            return (err, Vec::new());
+        }
+        if is_self_value(path, ns) {
```
Comment (Centril) [inline src/librustc_resolve/lib.rs]: Split out to a function that returns the `err`?

**Review by Centril — COMMENTED** [review]
Some potential places for refactoring... Not sure if the result will be better without looking at all dependencies, but just from the code structure, that's where I'd break into new functions.

**Comment by Centril** [comment]
I only looked briefly at the structure... I'll leave the actual review for r? @petrochenkov.

File: src/librustc_resolve/lib.rs (hunk 4 above)
Comment (alexreg) [reply src/librustc_resolve/lib.rs]: Yeah, I agree with this.

File: src/librustc_resolve/lib.rs (hunk 2 above)
Comment (alexreg) [reply src/librustc_resolve/lib.rs]: This is maybe a bit excessive to factor it out. Very @centril-ish. ;-)

File: src/librustc_resolve/lib.rs [hunk 10]
```diff
@@ -3123,391 +3123,417 @@ impl<'a> Resolver<'a> {
...
-                                   path: &[Segment],
-                                   span: Span,
-                                   source: PathSource,
-                                   crate_lint: CrateLint)
-                                   -> PathResolution {
-        let ident_span = path.last().map_or(span, |ident| ident.ident.span);
+    /// Provides context-dependent help for errors reported by the `smart_resolve_path_fragment`
+    /// function.
+    /// Returns `true` if able to provide context-dependent help.
+    fn smart_resolve_context_dep_help(
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]:
```suggestion
    fn smart_resolve_context_dependent_help(
```

**Comment by petrochenkov** [comment]
LGTM.
I'd prefer not to split the functions further.

Could you also move the factored out functions into `librustc_resolve/error_reporting.rs`?

**Comment by alexreg** [comment]
@petrochenkov Yeah I agree, further splitting is probably not optimal. We can even leave it at the double split (first commit only) if you prefer, but I'm happy with how it is now.

Points of review addressed.

**Comment by alexreg** [comment]
@Centril Definitely in need of one of your roll-up specials soon! ;-)

**Pull request merged**